crates/bevy_gilrs/src/rumble.rs

```diff
@@ -0,0 +1,146 @@
+//! Handle user specified rumble request events.
+use bevy_ecs::{
+    prelude::{EventReader, Res},
+    system::NonSendMut,
+};
+use bevy_input::gamepad::{GamepadRumbleIntensity, GamepadRumbleRequest};
+use bevy_log::{debug, warn};
+use bevy_time::Time;
+use bevy_utils::HashMap;
+use gilrs::{
+    ff::{self, BaseEffect, BaseEffectType},
+    GamepadId, Gilrs,
+};
+
+use crate::converter::convert_gamepad_id;
+
+struct RunningRumble {
+    deadline: f32,
+    // We use `effect.drop()` to interact with this, but rustc can't know
+    // gilrs uses Drop as an API feature.
+    #[allow(dead_code)]
+    effect: ff::Effect,
+}
+
+enum RumbleError {
+    GamepadNotFound,
+    GilrsError(ff::Error),
+}
+impl From<ff::Error> for RumbleError {
+    fn from(err: ff::Error) -> Self {
+        RumbleError::GilrsError(err)
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct RumblesManager {
+    rumbles: HashMap<GamepadId, Vec<RunningRumble>>,
+}
+
+fn to_gilrs_magnitude(ratio: f32) -> u16 {
+    (ratio * u16::MAX as f32) as u16
+}
+
+fn get_base_effects(
+    GamepadRumbleIntensity { weak, strong }: GamepadRumbleIntensity,
+) -> Vec<ff::BaseEffect> {
+    let mut effects = Vec::new();
+    if strong > 0. {
+        effects.push(BaseEffect {
+            kind: BaseEffectType::Strong {
+                magnitude: to_gilrs_magnitude(strong),
+            },
+            ..Default::default()
+        });
+    }
+    if weak > 0. {
+        effects.push(BaseEffect {
+            kind: BaseEffectType::Strong {
+                magnitude: to_gilrs_magnitude(weak),
+            },
+            ..Default::default()
+        });
+    }
+    effects
+}
+
+fn handle_rumble_request(
+    manager: &mut RumblesManager,
+    gilrs: &mut Gilrs,
+    rumble: GamepadRumbleRequest,
+    current_time: f32,
+) -> Result<(), RumbleError> {
+    let gamepad = match rumble {
+        GamepadRumbleRequest::Add { gamepad, .. } | GamepadRumbleRequest::Stop { gamepad } => {
+            gamepad
+        }
+    };
+    let (gamepad_id, _) = gilrs
+        .gamepads()
+        .find(|(pad_id, _)| convert_gamepad_id(*pad_id) == gamepad)
+        .ok_or(RumbleError::GamepadNotFound)?;
```

Comment:
hmm, so what happens when you `GamepadRumbleRequest::Add` with an already running request. Does gilrs documents it? You seem to add it to the list of rumbles for the given controller.

Supposedly you documented the behavior in the `GamepadRumbleRequest` doc. You did it in knowledge of the gilrs implementation right?

Reply:
Yes, gilrs does a saturating add of multiple effects, if i understood your question properly.

Reply:
I feel like this might be good to document this. It's intent that's not obviously expressed in the code, therefore warrants a comment. I suggest it as a doc string for the `RumblesManager::rumbles` field.